### PR TITLE
feat(ai): Gemini correctness discovery with repository scanner and prompt builder (#635)

### LINKED ISSUE #635: feat(ai): 定義 Gemini correctness discovery contract 與安全政策
Part of #634

## 目的

先建立**只讀 discovery 階段**：Gemini 可以掃描允許的程式碼與測試，但不能修改 production code、不能執行 shell、不能操作 GitHub。它每次只能輸出一個 strict JSON finding，或明確回報 `no-finding`。

此子票只負責「找出候選問題並把結果變成可驗證的機器合約」，不實作 RED 測試、GREEN 修復或 PR 發布。

## 前置調查

實作前先確認 #334、#378 已合併的 Gemini client／schema validation／secret 讀取方式，能共用就抽成小型 adapter，不要平行建立第二套互不相容的 Gemini 封裝。

需確認並記錄：

- 現有 Gemini SDK／REST client 放在哪裡。
- API key 的既有 secret 名稱。
- 是否已有 strict JSON／structured output helper。
- 是否已有 retry、quota、timeout、logging redaction。

## 建議檔案

實作者可依 repo 現況微調名稱，但責任必須分離：

```text
scripts/gemini-correctness/
  discover.mjs
  finding-schema.mjs
  policy.mjs
  prompts/discover.md
  __tests__/finding-schema.test.ts
```

不要把 policy、prompt、Gemini 呼叫與 GitHub workflow 全塞進單一 YAML heredoc。

## Discovery 輸入

只提供必要且不敏感的 repository 資訊：

- 目前 commit SHA。
- allowlisted source／test files。
- `CLAUDE.md` 中與 architecture、TDD、不變條件、內容可見性有關的規則。
- baseline 檢查摘要，不包含 secret、完整環境變數或 token。
- 可選的最近變更範圍，但不得讓 discovery 只看 git diff 而忽略既有邏輯。

第一版建議允許掃描：

- `src/domain/**`，但排除本票 protected paths。
- `src/hooks/**`。
- 純邏輯 helper。
- 對應 `*.test.ts`／`*.test.tsx`。

React 視覺、文案、題庫與 i18n 不列入自動 repair target。

## Finding schema

至少支援兩種結果。

### Finding

```json
{
  "schemaVersion": 1,
  "status": "finding",
  "title": "handle empty queue without dropping pending item",
  "confidence": 0.94,
  "category": "boundary-condition",
  "evidence": [
    {
      "file": "src/domain/example.ts",
      "startLine": 42,
      "endLine": 57,
      "reason": "The empty branch returns before preserving the pending item."
    }
  ],
  "expectedBehavior": "...",
  "actualBehavior": "...",
  "reproduction": {
    "testFile": "src/domain/example.regression.test.ts",
    "testName": "preserves pending item when queue becomes empty"
  },
  "productionFiles": ["src/domain/example.ts"],
  "risk": "low"
}
```

### No finding

```json
{
  "schemaVersion": 1,
  "status": "no-finding",
  "reason": "No safe high-confidence correctness issue found"
}
```

## Deterministic validation

不要只靠 prompt。validator 必須拒絕：

- 非 JSON、markdown fence、trailing prose。
- 未知欄位或缺必要欄位。
- `schemaVersion` 不支援。
- `confidence` 非有限數值、超出 0–1，或低於設定 threshold。
- `risk` 不是 `low`。
- evidence 沒有有效檔案／行號／理由。
- path traversal、絕對路徑、symlink escape。
- `productionFiles` 超出 allowlist 或命中 protected paths。
- reproduction test 不在 production file 同目錄，或不是 `*.regression.test.ts(x)`。
- finding 涉及題庫、翻譯、content visibility、auth、RLS、dependency、workflow、migration、generated file。
- 一次提出多個 bug，或 title／evidence 無法對應同一 root cause。

## Protected paths

policy 至少必須包含父票 #634 的限制，並用程式碼維護，而不是只有 prompt 文字：

```text
.github/**
.env*
package.json dependency versions
pnpm-lock.yaml
src/domain/contentGuard.ts
src/domain/types.ts contract-breaking changes
src/i18n.ts launched-language / fallback contract
src/domain/exam/items/**
scripts/exam-batches/**
src/domain/furiganaData.ts
Supabase migrations / RLS / grants
auth / billing / deployment / generated files
```

若 production file 位於 protected scope，結果應轉成 rejected finding，workflow 不得進入 RED。

## Prompt 要求

Discovery prompt 必須要求：

- 只找可由 deterministic unit／integration test 重現的行為錯誤。
- 優先 boundary condition、state transition、null／empty input、off-by-one、stale state、錯誤 fallback。
- 不提出 formatting、命名、效能猜測、dependency upgrade、架構重構。
- 不把缺測試本身視為 bug。
- 不修改任何檔案。
- 沒有高可信度 finding 時輸出 `no-finding`，不可硬湊。
- evidence 必須引用實際檔案與行號，不可捏造 symbol。

## Gemini adapter 安全要求

- API key 只從 process environment 讀取，不寫入 log／exception／artifact。
- model 名稱由 repository variable／CLI 參數傳入，不散落硬編碼。
- 設定 request timeout、有限重試與 exponential backoff。
- quota／429／5xx 與 invalid response 要有不同 exit code 或 machine-readable status。
- log 僅保留必要 metadata；不可輸出完整 request header。
- 測試使用 fake client，不呼叫真實 Gemini API。

## 測試

至少覆蓋：

- valid finding。
- valid no-finding。
- 低 confidence。
- 多 finding／混合 root cause。
- protected path。
- path traversal。
- 不存在檔案／越界行號。
- 非 regression test 命名。
- unknown field。
- malformed JSON／markdown wrapped JSON。
- quota／timeout／invalid structured response 的 adapter behavior。

## 驗收標準

- [ ] 有獨立 discovery runner、schema validator、policy 與 prompt。
- [ ] Discovery 不修改 repository 檔案。
- [ ] Finding／no-finding 皆為 strict JSON 並有 schema version。
- [ ] 每次最多一個 low-risk、高 confidence finding。
- [ ] protected paths／path traversal／不存在檔案會被 deterministic validator 阻擋。
- [ ] reproduction test path 由 validator 限定為 colocated `*.regression.test.ts(x)`。
- [ ] Gemini API key 不會出現在 stdout、stderr、artifact 或測試 snapshot。
- [ ] adapter 有 timeout、有限 retry、quota／invalid-response 診斷。
- [ ] validator 與 adapter 測試全綠。
- [ ] `pnpm typecheck`、`pnpm test`、`pnpm build` 通過。

## 明確不做

- 不修改程式碼或測試。
- 不執行任意 shell command。
- 不建立 branch／commit／PR。
- 不處理題庫、翻譯或日文內容正確性。
- 不允許 medium／high risk finding。

## 交接建議

Claude Code 先做 threat-model／contract review；Codex 依 contract 實作 validator 與 tests。PR body 使用 `Part of #634`，不要關閉父票。

### PR DESCRIPTION
**Pull request by nurockplayer**
Closes #635
Part of #634

## Summary

Implements the Gemini correctness discovery phase with full file scanning, prompt building, and dual-layer validation.

### scanner.mjs
- Reads allowlisted files (src/domain/**, src/hooks/**) with actual content and line numbers
- realpath-based symlink containment, hard limits (validated), UTF-8 byte truncation
- Accurate protectedExcluded counting, stable deterministic sort

### prompt-builder.mjs
- Builds prompt with COMPLETE file blocks only (no prompt.slice)
- True MAX_TOTAL_CHARS (500K) hard cap: computes full candidate prompt before accepting each block
- Single DEFAULT_MODEL (gemini-2.5-flash), strict JSON schema output instructions

### repo-validator.mjs — Repository-aware validation (second layer after JSON schema)
- Evidence files must exist, be regular files, inside repo, and in scanned manifest
- validateEvidenceLines uses visibleLineCount from scanner metadata (never from Gemini)
- Production files validated for existence, allowlist match, protected-path exclusion

### finding-schema.mjs — Pure JSON schema validation
- Rejects unknown fields, unsupported versions, out-of-range confidence, non-low risk, multi-root-cause

### gemini-client.mjs — Secure REST adapter
- Configurable timeout (>=1ms), finite retries (exponential backoff), 429/5xx handling
- API key redaction in all error paths (bodies, network errors, schema errors)
- Requires model option (no DEFAULT_MODEL — centralized to prompt-builder)

### policy.mjs — Canonical path security rules
- isProtected, isAllowlisted, isPathSafe, safeWritePath, isPathWithinRepo
- safeWritePath: creates .tmp if missing on clean checkout, rejects symlink escapes
- All modules import path rules from this single source

### discover.mjs — Orchestrator
- Reads project rules from REPO_ROOT/CLAUDE.md only (--claude-md removed, no arbitrary paths)
- realpath containment check on CLAUDE.md before reading
- Passes scannedFiles metadata to validator for visible line count lookup

## Tests

109 test files, 1191 tests — 0 failures.

## Verification

- pnpm typecheck — PASS
- pnpm test (109 files, 1191 tests) — 0 failures
- pnpm build — PASS
- git diff --check — PASS
- GitHub Actions CI — SUCCESS
- Codex review — SAFE

## Security

- `--claude-md` removed: CLAUDE.md read from canonical REPO_ROOT path with realpath containment
- All output constrained to .tmp/ via safeWritePath (symlink-aware)
- All API output deep-redacted for AIza... patterns
- Evidence files validated against scanned manifest (cannot reference unseen files)
- Line numbers validated against visible line count (truncated files cannot reference unseen lines)
- Model centralized to single DEFAULT_MODEL in prompt-builder.mjs

### CODE REVIEW

File: scripts/gemini-correctness/discover.mjs [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  return o;
+}
+
+// ---------------------------------------------------------------------------
+// Constrain output path to a tmp dir under REPO_ROOT/.tmp/
+// ---------------------------------------------------------------------------
+function resolveOutputPath(cliPath) {
+  if (!cliPath) return null;
+  const allowedDir = path.join(REPO_ROOT, ".tmp");
+  const safe = safeWritePath(cliPath, allowedDir);
```
[the remark this answers is not in the document]
Comment (nurockplayer) [reply scripts/gemini-correctness/discover.mjs]: 已由 commit 4bb8613 修復：safeWritePath 現在傳入 REPO_ROOT 作為第三參數。resolve.

File: scripts/gemini-correctness/policy.mjs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import path from "node:path";
+
+export function getDefaultProtectedPaths() {
+  return [
+    ".github/", ".env", ".env.", "pnpm-lock.yaml", "package.json",
+    "src/domain/contentGuard.ts", "src/domain/types.ts", "src/i18n.ts",
+    "src/domain/exam/items/", "scripts/exam-batches/",
+    "src/domain/furiganaData.ts", "supabase/", "functions/",
+    "src/domain/furiganaExplanationData.ts", "src/domain/furiganaLearningData.ts",
+    "src/domain/furigana", "src/domain/exam/examBlocks.ts",
```
[the remark this answers is not in the document]
Comment (nurockplayer) [reply scripts/gemini-correctness/policy.mjs]: 已由 commit 4bb8613 修復：policy.mjs 改為保護實際存在的 src/domain/examBlocks.ts。resolve.

**Review by nurockplayer — COMMENTED** [review]
Both P2 review comments resolved in commit 4bb8613. safeWritePath now passes REPO_ROOT; examBlocks path corrected to src/domain/examBlocks.ts.
